python/dgl/data/dgl_dataset.py

```diff
@@ -0,0 +1,157 @@
+            Use dgl.utils.data.load_graphs to load dgl graph from files.
+            Use dgl.utils.data.load_info to load extra information into python dict object.
+        """
+        pass
+
+    def process(self, root_path):
+        r"""Processes the data under root_path.
+            By default root_path = os.path.join(self.raw_dir, self.name).
+            One can overwrite raw_path() function to change the path.
+        """
+        raise NotImplementedError
+
+    def has_cache(self):
+        r"""Decide whether there exists a preprocessed dataset
+            By default False.
+        """
+        return False
+
+    @retry_method_with_fix(download)
+    def _download(self):
+        r"""Download dataset by calling self.download() if the dataset does not exists under self.raw_path.
+            By default self.raw_path = os.path.join(self.raw_dir, self.name)
+            One can overwrite raw_path() function to change the path.
+        """
+        if os.path.exists(self.raw_path):  # pragma: no cover
+            return
+
+        makedirs(self.raw_dir)
+        self.download()
+
+    def _load(self):
+        r"""Entry point from __init__ to load the dataset. 
+            if the cache exists:
+                Load the dataset from saved dgl graph and information files.
+            else:
+                1. Download the dataset if needed.
+                2. Process the dataset and build the dgl graph.
+                3. Save the processed dataset into files.
+        """
+        if not self._force_reload and self.has_cache():
+            self.load()
+        else:
+            self._download()
+            self.process(self.raw_path)
+            self.save()
+    
+    @property
+    def url(self):
+        r"""Get url to download the raw dataset
+        """
+        return self._url
+
+    @property
+    def name(self):
+        r"""Name of the dataset
+        """
+        return self._name
+
+    @property
+    def raw_dir(self):
+        r"""Raw file directory contains the input data directory.
+        """
+        return self._raw_dir
+
+    @property
+    def raw_path(self):
+        r"""File directory contains the input data.
+            By default raw_path = os.path.join(self.raw_dir, self.name)
+        """
+        return os.path.join(self.raw_dir, self.name)
+
+    def __getitem__(self, idx):
+        r"""Gets the data object at index.
+        """
+        raise NotImplementedError
+
+
+    def __len__(self):
+        r"""The number of examples in the dataset."""
+        raise NotImplementedError
+
+class DGLBuiltinDataset(DGLDataset):
+    r"""The Basic DGL Builtin Dataset.
+        Builtin dataset will be automatically downloaded into ~/.dgl/
+    Parameters
+    name : str
+        Name of the dataset
+    url : str
+        Url to download the raw dataset
+    force_reload : bool
+        Whether to reload the dataset. Default: False
+    """
+    def __init__(self, name, url, force_reload=False):
```

Comment:
Since `force_reload=False` here, I suppose all sub-classes in the API doc can get rid of it?

Reply:
Yes